gameobject.h

```diff
@@ -43,6 +43,11 @@ class GameObject
     bool CheckForCollision(GameObject& other);
 
     void Render() const;
+    template <typename T>
+    T* GetObject()
```

Comment:
This function is meant to get the actual derived gameobject class from an object. In general it would be something like
```
if (object->GetInstanceName("Ball"))
{
     Ball* ball = object->GetObject<Ball>();
     ball->SomeFunction();
}
```
or 
```
if (object->GetObject<Ball>() != nullptr)
{
    object->GetObject<Ball>()->SomeFunction();
}
```
to use the actual class in some way if needed, like in OnCollisionStart() for example. I know raw pointers are advised against, but I can't really return a reference here either? Becuase I need to be able to check if it is null if the object is not of this type. But I don't think I could return a shared_ptr here either, becuase that would conflict the shared_ptr in the Pong class' list of gameobject. Then it would cause a double release? What do you think? Is there a better way to do this in general?

Reply:
Raw pointers are totally good in the right circumstance :) As described in another comment, just avoid "raw new". I think the guidance is "use references whenever you can, and use pointers when you must". Thist may be one of those cases!!

In general, the options for returning a "Something or nothing" value in C++ is nullptr, so I think that's totally good. In the future, I think it will move more towards using std::optional to allow for that with value semantics.

> But I don't think I could return a shared_ptr here either, becuase that would conflict the shared_ptr in the Pong class' list of gameobject.

Yep, you shouldn't ever wrap something you didn't allocate manually in a shared pointer. You have good intuition there.

I think given what you've said about Unity doing it this way, this is fine for now! Have you peeked under the hood at Unity and confirmed they do it this way? Wonder if they have some kind of other tricky stuff up their sleeve